### PR TITLE
add traceable fetch

### PR DESCRIPTION
**Pull request by GrahamSH-LLK**
adds traceable fetch

### CODE REVIEW

**Comment by Samq64** [comment]
Could you add some context? I know this is related to #7648 but how does this actually prevent future problems?

**Comment by WorldLanguages** [comment]
I will be working on reducing the amount of requests to `scratch.mit.edu/session/` soon.

**Comment by CST1229** [comment]
> Could you add some context? I know this is related to #7648 but how does this actually prevent future problems?

i assume this means that if sa is making too many requests, the st can track down what exactly caused it

**Comment by GrahamSH-LLK** [comment]
> Could you add some context? I know this is related to #7648 but how does this actually prevent future problems?

Jeffalo has requested this so that Scratch can properly identify our traffic and work with us to not actively harm the site. We have over 280k users- requests we make can have _huge_ impact. See the discord for one such case. Even if we are fixing some of these issues, this is very helpful for any future requests we add, and for identifying additional pain points for load on Scratch. This change set is fairly minimal if you ignore the very simple (and not breaking) fetch additions to all userscripts.

**Comment by Hans5958** [comment]
The explanation should be put on the initial description, not in demand. At least #7646 should be mentioned as the initial PR of everything optimization.

File: libraries/common/cs/fetch.js
[the remark this answers is not in the document]
Comment (GrahamSH-LLK) [reply libraries/common/cs/fetch.js:1]: CORS. We can't just send arbitrary headers.

[the remark this answers is not in the document]
Comment (WorldLanguages) [reply libraries/common/cs/fetch.js:1]: This is a bit unfortunate, by adding our own headers we may be bypassing any existing user-side or server-side cache.

**Comment by WorldLanguages** [comment]
If we go forward with this, we may want to make this a little more clever. Adding URL params to all our requests may affect Scratch's ability to respond with cached resources, we would essentially become a separate caching namespace.

Maybe we can add some randomness, for example only 1% of users or requests sending a separate URL parameter. This way, 99% or our requests can return cached responses. And assuming our scale is big enough, multiplying the amount of SA requests by 100 would get very close to the actual amount of requests.

File: libraries/common/cs/fetch.js
[the remark this answers is not in the document]
Comment (DNin01) [reply libraries/common/cs/fetch.js:1]: Could we modify the user agent?

[the remark this answers is not in the document]
Comment (DNin01) [reply libraries/common/cs/fetch.js:1]: > CORS. We can't just send arbitrary headers.

I thought that having host permission (site access) bypassed CORS. Would that work here?

[the remark this answers is not in the document]
Comment (WorldLanguages) [reply libraries/common/cs/fetch.js:1]: > Could we modify the user agent?

We can, but I don't think the chrome.declarativeNetRequest API is powerful enough to let us distinguish our requests from vanilla requests, and we're back to square one. Unless we use a URL parameter to let DNR know, and somehow tell DNR to remove it before sending the request across the internet. (Probably impossible)

> I thought that having host permission (site access) bypassed CORS. Would that work here?

Yes, but adding extra headers is a separate type of CORS related to sending OPTIONS headers, and we would be sending twice the requests for every request we make. Also, we can only bypass CORS from the background process of the extension, and addon requests don't go through there.

**Comment by WorldLanguages** [comment]
> Maybe we can add some randomness, for example only 1% of users or requests sending a separate URL parameter.

Actually, this misses the point. Adding a URL parameter to all requests would allow Scratch to block all requests before they reach their infrastructure, effectively acting as a kill switch.

**Comment by WorldLanguages** [comment]
I find it unlikely we'll add URL parameters to all of our requests at once, though. The scope of this PR is too big.

What I'm thinking is that we should rethink the code that previously caused an infinite loop (handle-auth.js) and add URL parameters there so that if our new implementation ends up being bad, it can be blocked by Scratch. That would basically remove the risk of making things worse. We can then add an identifying URL parameter to all other requests, starting with other background code, and finally addons which run in Scratch tabs.

**Comment by WorldLanguages** [comment]
The communication here is a bit awkward, some things are only said on Discord, and some Scratch comments I'm only finding now, for example: https://scratch.mit.edu/users/codubee/#comments-338105607

But, as a summary, these are the things I am sure about:
- Adding a header or modifying User-Agent is too complicated, or maybe impossible.
- It wouldn't be problematic for Scratch if we add a `?scratchaddons=1` parameter to all of our requests.
- It would be helpful both to measure the impact of our requests, as well as blocking them if necessary. Most automated requests, for example those sent by non-browsers, are easily identifiable due to different User-Agent headers and some missing headers such as those that start with "Sec-". Requests sent by extensions are really hard (wouldn't say impossible, it depends) to distinguish from normal traffic, which is problematic.
- When making any changes to code that sends requests to Scratch servers, we should always add the parameter so that the requests can be blocked if necessary.
- I don't think the approach this PR takes is a good idea, I would prefer to add our URL parameter more gradually and carefully.

Okay. To move on, I think we should continue with our work to send less requests and avoid bugs, and add our URL parameter, both at the same time.

Specifically for v1.39.0, I suggest we focus on the following changes:
- We should fix the root cause of the infinite loop. That is, we should avoid sending **any** requests in response to `chrome.cookies.onChanged`.
- For any changes related to the previous point, we should add URL parameters to any requests we send, for example, `?sadd=1&saddinfo=bg`.
- We could add URL parameters to most/all requests sent by the background process of the extension.
- Specifically when fetching message counts in the background process, we could use `(await chrome.action.getUserSettings()).isOnToolbar` to detect whether the user has our extension pinned in the extension toolbar. If not, it makes very little sense to periodically obtain the user's message count.
- We need to recognize that most of our modern users aren't that interested on Scratch Messaging, yet we enable those features by default due to historical reasons. Ironically, messaging (as well as `msg-count-badge`) is one of the features that requests data periodically from Scratch servers.

**Pull request closed without merging**